README.md
# VAT.fyi SDK PHP

## Installing SDK

The recommended way to install SDK is through
[Composer](http://getcomposer.org).

```bash
# Install Composer
curl -sS https://getcomposer.org/installer | php
```

Next, run the Composer command to install the latest stable version of SDK:

```bash
php composer.phar require richweber-technology/vatfyi-sdk-php
```

After installing, you need to require Composer's autoloader:

```php
require 'vendor/autoload.php';
```

You can then later update SDK using composer:

 ```bash
php composer.phar update
 ```

## Use

```php
use RichweberTechnology\vatfyi\Client;
use RichweberTechnology\vatfyi\components\Company;
use RichweberTechnology\vatfyi\components\VatNumber;

$client = new Client('YOUR_API_KEY');
$checker = new VatNumber($client);

$check = $checker->checkVatNumber('EE102001059');

echo $check . PHP_EOL;
echo $checker->isSuccess() . PHP_EOL;
echo $checker->isValidNumber() . PHP_EOL;
echo $checker->getErrorDescription() . PHP_EOL;

$company = $checker->getCompany();
if ($company instanceof Company) {
    echo $company->getCompanyName() . PHP_EOL;
    echo $company->getCountryCode() . PHP_EOL;
    echo $company->getCompanyAddress() . PHP_EOL;
    echo $company->getVatNumber() . PHP_EOL;
}
```

```php
use RichweberTechnology\vatfyi\Client;
use RichweberTechnology\vatfyi\components\Amount;
use RichweberTechnology\vatfyi\components\RateDto;
use RichweberTechnology\vatfyi\components\VatRate;

$client = new Client('YOUR_API_KEY');

$dto = new RateDto('EE', 123.45);
$checker = new VatRate($client);

$check = $checker->getVatRate($dto);

echo $check . PHP_EOL;
echo $checker->isSuccess() . PHP_EOL;
echo $checker->isVatNumberConfirmed() . PHP_EOL;
echo $checker->isVatNumberFailed() . PHP_EOL;
echo $checker->getVatNumberFailDescription() . PHP_EOL;
echo $checker->getErrorDescription() . PHP_EOL;

$company = $checker->getAmount();
if ($company instanceof Amount) {
    echo $company->getVAT() . PHP_EOL;
    echo $company->getAmount() . PHP_EOL;
    echo $company->getVatAmount() . PHP_EOL;
    echo $company->getTotalAmount() . PHP_EOL;
}
```


CHANGELOG.md
# VAT.fyi SDK PHP Change Log

### 1.0.3 November 18, 2022

- Upgraded to PHP 8.2

### 1.0.2 November 27, 2021

- Upgraded to PHP 8.1

### 1.0.1 December 09, 2020

- Upgraded to PHP 8

### 1.0.0 January 15, 2019

- Created extension


src/Client.php
<?php

namespace RichweberTechnology\vatfyi;

use GuzzleHttp\Client as GuzzleClient;

/**
 * Class Client
 * @package RichweberTechnology\vatfyi
 */
class Client
{
    /**
     * @var string
     */
    private $_serviceURL = 'https://api.vat.fyi';

    /**
     * @var string
     */
    private $_apiVersion = '1';

    /**
     * @var string
     */
    private $_apiKey;

    /**
     * @var GuzzleClient
     */
    private $_guzzle;

    /**
     * @var bool
     */
    private $_isTest = false;

    /**
     * @var bool
     */
    private $_returnRawResponse = true;

    /**
     * Client constructor.
     *
     * @param string $apiKey
     * @param array $params
     */
    public function __construct(string $apiKey = '', array $params = [])
    {
        $this->_apiKey = $apiKey;
        $this->_guzzle = new GuzzleClient();

        if (isset($params['returnRawResponse']) && is_bool($params['returnRawResponse'])) {
            $this->_returnRawResponse = $params['returnRawResponse'];
        }
        if (isset($params['isTest']) && is_bool($params['isTest'])) {
            $this->_isTest = $params['isTest'];
        }
        if (isset($params['serviceURL'])) {
            $this->_serviceURL = $params['serviceURL'];
        }
        if (isset($params['apiVersion'])) {
            $this->_apiVersion = $params['apiVersion'];
        }
    }

    /**
     * @param string $serviceURL
     *
     * @return Client
     */
    public function setServiceURL(string $serviceURL)
    {
        $this->_serviceURL = $serviceURL;
        return $this;
    }

    /**
     * @param string $apiVersion
     *
     * @return Client
     */
    public function setApiVersion(string $apiVersion)
    {
        $this->_apiVersion = $apiVersion;
        return $this;
    }

    /**
     * @param bool $isTest
     *
     * @return Client
     */
    public function setIsTest(bool $isTest)
    {
        $this->_isTest = $isTest;
        return $this;
    }

    /**
     * @param bool $returnRawResponse
     *
     * @return Client
     */
    public function setReturnRawResponse(bool $returnRawResponse)
    {
        $this->_returnRawResponse = $returnRawResponse;
        return $this;
    }

    /**
     * @param string $requestType
     * @param string $method
     * @param array $data
     *
     * @return string
     * @throws \GuzzleHttp\Exception\GuzzleException
     */
    public function getResponse(string $requestType, string $method, array $data = [])
    {
        $response = $this->_guzzle->request($requestType, $this->getURL($method), [
            'http_errors' => !$this->_returnRawResponse,
            'verify' => !$this->_isTest,
            'query' => ['access-token' => $this->_apiKey],
            'json' => $data,
            'headers' => [
                'Accept' => 'application/json',
            ],
        ]);

        return (string)$response->getBody();
    }

    /**
     * @param string $method
     *
     * @return string
     */
    private function getURL(string $method)
    {
        return $this->_serviceURL . '/v' . $this->_apiVersion . '/' . $method;
    }
}


src/components/Amount.php
<?php

namespace RichweberTechnology\vatfyi\components;

/**
 * Class Amount
 * @package RichweberTechnology\vatfyi\components
 */
class Amount
{
    /**
     * @var string
     */
    private $_VAT;

    /**
     * @var float
     */
    private $_amount;

    /**
     * @var float
     */
    private $_vatAmount;

    /**
     * @var float
     */
    private $_totalAmount;

    /**
     * Amount constructor.
     *
     * @param string $VAT
     * @param float $amount
     * @param float $vatAmount
     * @param float $totalAmount
     */
    public function __construct(string $VAT, float $amount, float $vatAmount, float $totalAmount)
    {
        $this->_VAT = $VAT;
        $this->_amount = $amount;
        $this->_vatAmount = $vatAmount;
        $this->_totalAmount = $totalAmount;
    }

    /**
     * @return string
     */
    public function getVAT(): string
    {
        return $this->_VAT;
    }

    /**
     * @return float
     */
    public function getAmount(): float
    {
        return $this->_amount;
    }

    /**
     * @return float
     */
    public function getVatAmount(): float
    {
        return $this->_vatAmount;
    }

    /**
     * @return float
     */
    public function getTotalAmount(): float
    {
        return $this->_totalAmount;
    }
}


src/components/Company.php
<?php

namespace RichweberTechnology\vatfyi\components;

/**
 * Class Company
 * @package RichweberTechnology\vatfyi\components
 */
class Company
{
    /**
     * @var string
     */
    private $_companyName;

    /**
     * @var string
     */
    private $_countryCode;

    /**
     * @var string|null
     */
    private $_companyAddress;

    /**
     * @var string|null
     */
    private $_vatNumber;

    /**
     * Company constructor.
     *
     * @param string $companyName
     * @param string $countryCode
     * @param string|null $companyAddress
     * @param string|null $vatNumber
     */
    public function __construct(
        string $companyName,
        string $countryCode,
        ?string $companyAddress = null,
        ?string $vatNumber = null
    ) {
        $this->_companyName = $companyName;
        $this->_countryCode = $countryCode;
        $this->_companyAddress = $companyAddress;
        $this->_vatNumber = $vatNumber;
    }

    /**
     * @return string
     */
    public function getCompanyName(): string
    {
        return $this->_companyName;
    }

    /**
     * @return string
     */
    public function getCountryCode(): string
    {
        return $this->_countryCode;
    }

    /**
     * @return string|null
     */
    public function getCompanyAddress(): ?string
    {
        return $this->_companyAddress;
    }

    /**
     * @return string|null
     */
    public function getVatNumber(): ?string
    {
        return $this->_vatNumber;
    }
}


src/components/RateDto.php
<?php

namespace RichweberTechnology\vatfyi\components;

use RichweberTechnology\vatfyi\VatException;

/**
 * Class RateDto
 * @package RichweberTechnology\vatfyi\components
 */
class RateDto
{
    /**
     * @var string
     */
    private $_clientCountryCode;

    /**
     * @var float
     */
    private $_amount;

    /**
     * @var boolean
     */
    private $_isBusinessClient = false;

    /**
     * @var boolean
     */
    private $_hasVatNumber = false;

    /**
     * @var string|null
     */
    private $_vatNumber;

    /**
     * RateDto constructor.
     *
     * @param string $clientCountryCode
     * @param float $amount
     * @param bool $isBusinessClient
     * @param bool $hasVatNumber
     * @param string|null $vatNumber
     *
     * @throws VatException
     */
    public function __construct(
        string $clientCountryCode,
        float $amount,
        bool $isBusinessClient = false,
        bool $hasVatNumber = false,
        $vatNumber = null
    ) {
        $this->_clientCountryCode = $clientCountryCode;
        $this->_amount = $amount;

        if ($isBusinessClient && $hasVatNumber && (empty($vatNumber) || $vatNumber === '')) {
            throw new VatException('VAT number can\'t be empty.');
        }
        if ($amount < 0.01) {
            throw new VatException('Amount should be more then 0.');
        }

        $this->_isBusinessClient = $isBusinessClient;
        $this->_hasVatNumber = $hasVatNumber;
        $this->_vatNumber = $vatNumber;
    }

    /**
     * @return string
     */
    public function getClientCountryCode()
    {
        return $this->_clientCountryCode;
    }

    /**
     * @return float
     */
    public function getAmount()
    {
        return $this->_amount;
    }

    /**
     * @return bool
     */
    public function isBusinessClient()
    {
        return $this->_isBusinessClient;
    }

    /**
     * @return bool
     */
    public function hasVatNumber()
    {
        return $this->_hasVatNumber;
    }

    /**
     * @return string|null
     */
    public function getVatNumber()
    {
        return $this->_vatNumber;
    }
}


src/components/VatNumber.php
<?php

namespace RichweberTechnology\vatfyi\components;

use RichweberTechnology\vatfyi\Client;

/**
 * Class VatNumber
 * @package RichweberTechnology\vatfyi\components
 */
class VatNumber
{
    /**
     * @var Client
     */
    private $_client;

    /**
     * @var Company|null
     */
    private $_company;

    /**
     * @var bool
     */
    private $_isValidNumber = false;

    /**
     * @var string
     */
    private $_errorDescription = '';

    /**
     * VatNumber constructor.
     *
     * @param Client $client
     */
    public function __construct(Client $client)
    {
        $this->_client = $client;
    }

    /**
     * @return bool
     */
    public function isSuccess()
    {
        return $this->_errorDescription === '';
    }

    /**
     * @param string $number
     *
     * @return bool
     */
    public function checkVatNumber(string $number)
    {
        try {
            $response = $this->_client->getResponse('GET', 'vat/number/' . $number . '/check');
            $vat = json_decode($response);

            if ((int)$vat->status === 200) {
                if ($vat->data->isValidNumber) {
                    $this->_isValidNumber = true;
                    $this->_company = new Company(
                        $vat->data->company->name,
                        substr($number, 0, 2),
                        $vat->data->company->address,
                        substr($number, 2)
                    );
                }

                return true;
            }
            if ((int)$vat->status === 422) {
                foreach ($vat->data as $error) {
                    $this->_errorDescription .= $error->message . PHP_EOL;
                }
            }
            if ((int)$vat->status === 401) {
                $this->_errorDescription = $vat->data->message;
            }
        } catch (\GuzzleHttp\Exception\GuzzleException | \Exception $exception) {
            $this->_errorDescription = $exception->getMessage();
        }

        return false;
    }

    /**
     * @return Company|null
     */
    public function getCompany()
    {
        return $this->_company;
    }

    /**
     * @return bool
     */
    public function isValidNumber()
    {
        return $this->_isValidNumber;
    }

    /**
     * @return string
     */
    public function getErrorDescription()
    {
        return $this->_errorDescription;
    }
}


src/components/VatRate.php
<?php

namespace RichweberTechnology\vatfyi\components;

use RichweberTechnology\vatfyi\Client;

/**
 * Class VatRate
 * @package RichweberTechnology\vatfyi\components
 */
class VatRate
{
    /**
     * @var Client
     */
    private $_client;

    /**
     * @var Amount|null
     */
    private $_amount;

    /**
     * @var bool
     */
    private $_isVatNumberConfirmed = false;

    /**
     * @var bool
     */
    private $_isVatNumberFailed = false;

    /**
     * @var string|null
     */
    private $_vatNumberFailDescription;

    /**
     * @var string
     */
    private $_errorDescription = '';

    /**
     * VatNumber constructor.
     *
     * @param Client $client
     */
    public function __construct(Client $client)
    {
        $this->_client = $client;
    }

    /**
     * @return bool
     */
    public function isSuccess()
    {
        return $this->_errorDescription === '';
    }

    /**
     * @param RateDto $dto
     *
     * @return bool
     */
    public function getVatRate(RateDto $dto)
    {
        try {
            $data = [
                'clientCountryCode' => $dto->getClientCountryCode(),
                'amount' => $dto->getAmount(),
            ];

            if ($dto->isBusinessClient()) {
                $data['isBusinessClient'] = $dto->isBusinessClient();
            }
            if ($dto->hasVatNumber()) {
                $data['hasVatNumber'] = $dto->hasVatNumber();
            }
            if ($dto->getVatNumber()) {
                $data['vatNumber'] = $dto->getVatNumber();
            }

            $response = $this->_client->getResponse('POST', 'vat/rate/calculate', $data);
            $vat = json_decode($response);

            if ((int)$vat->status === 201) {
                $this->_amount = new Amount(
                    $vat->data->calculatedAmount->VAT,
                    (float)$vat->data->calculatedAmount->amount,
                    (float)$vat->data->calculatedAmount->vatAmount,
                    (float)$vat->data->calculatedAmount->totalAmount
                );

                $this->_isVatNumberConfirmed = $vat->data->isVatNumberConfirmed;
                $this->_isVatNumberFailed = $vat->data->isVatNumberFailed;
                $this->_vatNumberFailDescription = $vat->data->vatNumberFailDescription;

                return true;
            }
            if ((int)$vat->status === 422) {
                foreach ($vat->data as $error) {
                    $this->_errorDescription .= $error->message . PHP_EOL;
                }
            }
            if ((int)$vat->status === 401) {
                $this->_errorDescription = $vat->data->message;
            }
        } catch (\GuzzleHttp\Exception\GuzzleException | \Exception $exception) {
            $this->_errorDescription = $exception->getMessage();
        }

        return false;
    }

    /**
     * @return Amount|null
     */
    public function getAmount()
    {
        return $this->_amount;
    }

    /**
     * @return bool
     */
    public function isVatNumberConfirmed()
    {
        return $this->_isVatNumberConfirmed;
    }

    /**
     * @return bool
     */
    public function isVatNumberFailed()
    {
        return $this->_isVatNumberFailed;
    }

    /**
     * @return string|null
     */
    public function getVatNumberFailDescription()
    {
        return $this->_vatNumberFailDescription;
    }

    /**
     * @return string
     */
    public function getErrorDescription()
    {
        return $this->_errorDescription;
    }
}
